README.md
# kvmd-debian

PiKVM on Debian, Ubuntu or Armbian

## How to use

```bash
# TAG=$(curl "https://api.github.com/repos/hzyitc/kvmd-debian/releases/latest" | jq -r .tag_name)
TAG=debian-v3.142
VERSION=$(echo "$TAG" | grep -oE '[0-9]+([\.-][0-9]+)+')

# Donwload the packages
curl -L -O "https://github.com/hzyitc/kvmd-debian/releases/download/${TAG}/{python3-kvmd_${VERSION}_all.deb,kvmd-platform-v2-hdmiusb-generic_${VERSION}_all.deb}"

# Install them
dpkg -i python3-kvmd_${VERSION}_all.deb
dpkg -i kvmd-platform-v2-hdmiusb-generic_${VERSION}_all.deb
apt install --fix-broken --yes

# Prepare MSD storage
# a. Use a new partition to storage
#	mkfs.ext4 /dev/sdX
#	echo "LABEL=PIMSD  /var/lib/kvmd/msd  ext4  defaults,X-kvmd.otgmsd-user=kvmd  0  0" >> /etc/fstab
#	mount -a
#	mkdir -p /var/lib/kvmd/msd/{images,meta}
#	chown kvmd -R /var/lib/kvmd/msd/
# b. Storage in root partititon
sed -i -E 's/^([ \t]*)main\(\)$/\1#main()\n\1pass/' /usr/bin/kvmd-helper-otgmsd-remount
mkdir -p /var/lib/kvmd/msd/{images,meta}
chown kvmd -R /var/lib/kvmd/msd/

# Disable nginx to free http and https port
systemctl disable nginx

# Enable kvmd services
systemctl enable kvmd-otg kvmd-nginx kvmd

# Reboot and enjoy
reboot
```


debian/changelog
kvmd (3.142) unstable; urgency=low

  * Bump version: 3.141 → 3.142
  * pikvm/pikvm#813: optional page close confirmation

 -- Maxim Devaev <mdevaev@gmail.com>  Thu, 1 Sep 2022 12:24:58 +0300

kvmd (3.141) unstable; urgency=low

  * Bump version: 3.140 → 3.141
  * edidconf: fixed checksums update in write_bin()
  * lint fix

 -- Maxim Devaev <mdevaev@gmail.com>  Tue, 30 Aug 2022 15:55:23 +0300

kvmd (3.140) unstable; urgency=low

  * Bump version: 3.139 → 3.140
  * ocr: close old image on resizing
  * Zoom and Enhance before reading text (#113)

 -- Maxim Devaev <mdevaev@gmail.com>  Tue, 30 Aug 2022 05:53:39 +0300

kvmd (3.139) unstable; urgency=low

  * Bump version: 3.138 → 3.139
  * loc-bigpost.conf
  * removed body limit for /api/hid/print
  * new logo

 -- Maxim Devaev <mdevaev@gmail.com>  Sat, 27 Aug 2022 03:03:09 +0300

kvmd (3.138) unstable; urgency=low

  * Bump version: 3.137 → 3.138
  * pikvm/pikvm#375: fixed AltGr handling
  * refactoring
  * system menu min width
  * spoiler for mouse settings

 -- Maxim Devaev <mdevaev@gmail.com>  Mon, 22 Aug 2022 19:07:14 +0300

kvmd (3.137) unstable; urgency=low

  * Bump version: 3.136 → 3.137
  * cmdret module instead of cmd.read implementation

 -- Maxim Devaev <mdevaev@gmail.com>  Fri, 19 Aug 2022 21:47:46 +0300

kvmd (3.136) unstable; urgency=low

  * Bump version: 3.135 → 3.136
  * minor fix
  * add the ability to use command exit status as a gpio state in ugpio/cmd (#112)

 -- Maxim Devaev <mdevaev@gmail.com>  Wed, 17 Aug 2022 15:03:21 +0300

kvmd (3.135) unstable; urgency=low

  * Bump version: 3.134 → 3.135
  * /api/msd/read: zstd supported
  * refactoring

 -- Maxim Devaev <mdevaev@gmail.com>  Tue, 16 Aug 2022 18:59:18 +0300

kvmd (3.134) unstable; urgency=low

  * Bump version: 3.133 → 3.134
  * using shield_fg() in atomic (now atomic_fg)
  * using shield_fg()
  * deprecated all before 3.10
  * simplified AioNotifier()
  * aiotools.shield_fg()
  * compressed mode for /api/msd/read
  * new sponsors
  * hr fix
  * shielded some tasks
  * shield some finally ops
  * refactored msd writer api
  * refactored msd reader api

 -- Maxim Devaev <mdevaev@gmail.com>  Sun, 7 Aug 2022 19:43:00 +0300

kvmd (3.133) unstable; urgency=low

  * Bump version: 3.132 → 3.133
  * fixed window-button-exit-full-tab

 -- Maxim Devaev <mdevaev@gmail.com>  Mon, 1 Aug 2022 19:55:46 +0300

kvmd (3.132) unstable; urgency=low

  * Bump version: 3.131 → 3.132
  * pikvm/pikvm#631: option to hide navbar
  * new sponsor

 -- Maxim Devaev <mdevaev@gmail.com>  Mon, 1 Aug 2022 18:40:09 +0300

kvmd (3.131) unstable; urgency=low

  * Bump version: 3.130 → 3.131
  * Fixed pikvm/pikvm#783: remove incomplete msd image

 -- Maxim Devaev <mdevaev@gmail.com>  Sat, 30 Jul 2022 19:43:01 +0300

kvmd (3.130) unstable; urgency=low

  * Bump version: 3.129 → 3.130
  * cast fix
  * refactoring

 -- Maxim Devaev <mdevaev@gmail.com>  Sat, 30 Jul 2022 00:04:34 +0300

kvmd (3.129) unstable; urgency=low

  * Bump version: 3.128 → 3.129
  * common usb-keymap.h
  * new sponsor

 -- Maxim Devaev <mdevaev@gmail.com>  Fri, 29 Jul 2022 15:15:20 +0300

kvmd (3.128) unstable; urgency=low

  * Bump version: 3.127 → 3.128
  * fix
  * pikvm/pikvm#785: mute chown

 -- Maxim Devaev <mdevaev@gmail.com>  Tue, 26 Jul 2022 19:57:39 +0300

kvmd (3.127) unstable; urgency=low

  * Bump version: 3.126 → 3.127
  * new sponsors
  * better wording

 -- Maxim Devaev <mdevaev@gmail.com>  Mon, 25 Jul 2022 05:20:31 +0300

kvmd (3.126) unstable; urgency=low

  * Bump version: 3.125 → 3.126
  * report about downloading
  * msd: downloading status api
  * refactoring
  * improved msd rw ui
  * loc-nobuffering.conf
  * disabled buffering for /api/msd/read
  * fixed lock behaviour in read_image()

 -- Maxim Devaev <mdevaev@gmail.com>  Sun, 24 Jul 2022 18:08:40 +0300

kvmd (3.125) unstable; urgency=low

  * Bump version: 3.124 → 3.125

 -- Maxim Devaev <mdevaev@gmail.com>  Sun, 24 Jul 2022 06:02:55 +0300

kvmd (3.124) unstable; urgency=low

  * Bump version: 3.123 → 3.124
  * remount-ro on remove() is not fatal
  * process cdrom/rw forbidden states in set_params()
  * rw warning
  * reading images api
  * refactoring
  * upload_chunk_size -> write_chunk_size
  * renamed new_writer to writer
  * writable msd
  * refactoring
  * basic msd rw api
  * spell
  * clarify
  * icon instead of text
  * bold windows close icon
  * removed some newlines
  * refactoring
  * new sponsor
  * stm32 hid cleanup
  * avr hid: removed KEY_ERROR_UNDEFINED
  * Stm32 init (#100)
  * lint fix
  * type fix
  * using usb codes instead of arduino cinstants
  * renamed otg key to usb key
  * csv fix

 -- Maxim Devaev <mdevaev@gmail.com>  Sun, 24 Jul 2022 06:01:47 +0300

kvmd (3.123) unstable; urgency=low

  * Bump version: 3.122 → 3.123
  * new sponsor
  * fixed make upload
  * improved vnc logging
  * improved vnc encodings handling
  * refactoring
  * Support for driving TESmart KVMs by RS-232 (#89)

 -- Maxim Devaev <mdevaev@gmail.com>  Thu, 21 Jul 2022 17:30:14 +0300

kvmd (3.122) unstable; urgency=low

  * Bump version: 3.121 → 3.122
  * option to disable vencrypt
  * lint fixes
  * common event parsing
  * handle closing ws event
  * verbose rfb encdings report
  * log start_tls()
  * spell
  * lint fixes
  * improved aiotools.run()
  * workaround for https://bugs.python.org/issue39758
  * more vnc debug

 -- Maxim Devaev <mdevaev@gmail.com>  Wed, 20 Jul 2022 16:39:09 +0300

kvmd (3.121) unstable; urgency=low

  * Bump version: 3.120 → 3.121
  * hanged EDID monitor name to PiKVM and manufacturer to LNX

 -- Maxim Devaev <mdevaev@gmail.com>  Sat, 16 Jul 2022 23:14:31 +0300

kvmd (3.120) unstable; urgency=low

  * Bump version: 3.119 → 3.120
  * vulture fix
  * new sponsors
  * fixed make run
  * fixed tests
  * commented playoutDelay
  * Update outputs.h (#106)
  * playoutDelayHint = 0
  * refactoring
  * renamed patch (#103)
  * renamed patch
  * commented kvmd/kvmd#80 (#102)
  * commented kvmd/kvmd#80

kvmd/htserver.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2022  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


import os
import socket
import asyncio
import contextlib
import dataclasses
import inspect
import urllib.parse
import json

from typing import Tuple
from typing import List
from typing import Dict
from typing import Callable
from typing import AsyncGenerator
from typing import Optional
from typing import Union
from typing import Any

from aiohttp import ClientWebSocketResponse
from aiohttp.web import BaseRequest
from aiohttp.web import Request
from aiohttp.web import Response
from aiohttp.web import StreamResponse
from aiohttp.web import WebSocketResponse
from aiohttp.web import WSMsgType
from aiohttp.web import Application
from aiohttp.web import run_app
from aiohttp.web import normalize_path_middleware

try:
    from aiohttp.web import AccessLogger  # type: ignore
except ImportError:
    from aiohttp.helpers import AccessLogger  # type: ignore

from .logging import get_logger

from .errors import OperationError
from .errors import IsBusyError

from .validators import ValidatorError

from . import aiotools


# =====
class HttpError(Exception):
    def __init__(self, msg: str, status: int) -> None:
        super().__init__(msg)
        self.status = status


class UnauthorizedError(HttpError):
    def __init__(self) -> None:
        super().__init__("Unauthorized", 401)


class ForbiddenError(HttpError):
    def __init__(self) -> None:
        super().__init__("Forbidden", 403)


class UnavailableError(HttpError):
    def __init__(self) -> None:
        super().__init__("Service Unavailable", 503)


# =====
@dataclasses.dataclass(frozen=True)
class HttpExposed:
    method: str
    path: str
    auth_required: bool
    handler: Callable


_HTTP_EXPOSED = "_http_exposed"
_HTTP_METHOD = "_http_method"
_HTTP_PATH = "_http_path"
_HTTP_AUTH_REQUIRED = "_http_auth_required"


def exposed_http(http_method: str, path: str, auth_required: bool=True) -> Callable:
    def set_attrs(handler: Callable) -> Callable:
        setattr(handler, _HTTP_EXPOSED, True)
        setattr(handler, _HTTP_METHOD, http_method)
        setattr(handler, _HTTP_PATH, path)
        setattr(handler, _HTTP_AUTH_REQUIRED, auth_required)
        return handler
    return set_attrs


def _get_exposed_http(obj: object) -> List[HttpExposed]:
    return [
        HttpExposed(
            method=getattr(handler, _HTTP_METHOD),
            path=getattr(handler, _HTTP_PATH),
            auth_required=getattr(handler, _HTTP_AUTH_REQUIRED),
            handler=handler,
        )
        for handler in [getattr(obj, name) for name in dir(obj)]
        if inspect.ismethod(handler) and getattr(handler, _HTTP_EXPOSED, False)
    ]


# =====
@dataclasses.dataclass(frozen=True)
class WsExposed:
    event_type: str
    handler: Callable


_WS_EXPOSED = "_ws_exposed"
_WS_EVENT_TYPE = "_ws_event_type"


def exposed_ws(event_type: str) -> Callable:
    def set_attrs(handler: Callable) -> Callable:
        setattr(handler, _WS_EXPOSED, True)
        setattr(handler, _WS_EVENT_TYPE, event_type)
        return handler
    return set_attrs


def _get_exposed_ws(obj: object) -> List[WsExposed]:
    return [
        WsExposed(
            event_type=getattr(handler, _WS_EVENT_TYPE),
            handler=handler,
        )
        for handler in [getattr(obj, name) for name in dir(obj)]
        if inspect.ismethod(handler) and getattr(handler, _WS_EXPOSED, False)
    ]


# =====
def make_json_response(
    result: Optional[Dict]=None,
    status: int=200,
    set_cookies: Optional[Dict[str, str]]=None,
    wrap_result: bool=True,
) -> Response:

    response = Response(
        text=json.dumps(({
            "ok": (status == 200),
            "result": (result or {}),
        } if wrap_result else result), sort_keys=True, indent=4),
        status=status,
        content_type="application/json",
    )
    if set_cookies:
        for (key, value) in set_cookies.items():
            response.set_cookie(key, value)
    return response


def make_json_exception(err: Exception, status: Optional[int]=None) -> Response:
    name = type(err).__name__
    msg = str(err)
    if isinstance(err, HttpError):
        status = err.status
    else:
        get_logger().error("API error: %s: %s", name, msg)
    assert status is not None, err
    return make_json_response({
        "error": name,
        "error_msg": msg,
    }, status=status)


async def start_streaming(
    request: Request,
    content_type: str,
    content_length: int=-1,
    file_name: str="",
) -> StreamResponse:

    response = StreamResponse(status=200, reason="OK")
    response.content_type = content_type
    if content_length >= 0:
        response.content_length = content_length
    if file_name:
        file_name = urllib.parse.quote(file_name, safe="")
        response.headers["Content-Disposition"] = f"attachment; filename*=UTF-8''{file_name}"
    await response.prepare(request)
    return response


async def stream_json(response: StreamResponse, result: Dict, ok: bool=True) -> None:
    await response.write(json.dumps({
        "ok": ok,
        "result": result,
    }).encode("utf-8") + b"\r\n")


async def stream_json_exception(response: StreamResponse, err: Exception) -> None:
    name = type(err).__name__
    msg = str(err)
    get_logger().error("API error: %s: %s", name, msg)
    await stream_json(response, {
        "error": name,
        "error_msg": msg,
    }, False)


async def send_ws_event(
    wsr: Union[ClientWebSocketResponse, WebSocketResponse],
    event_type: str,
    event: Optional[Dict],
) -> None:

    await wsr.send_str(json.dumps({
        "event_type": event_type,
        "event": event,
    }))


def parse_ws_event(msg: str) -> Tuple[str, Dict]:
    data = json.loads(msg)
    if not isinstance(data, dict):
        raise RuntimeError("Top-level event structure is not a dict")
    event_type = data.get("event_type")
    if not isinstance(event_type, str):
        raise RuntimeError("event_type must be a string")
    event = data["event"]
    if not isinstance(event, dict):
        raise RuntimeError("event must be a dict")
    return (event_type, event)


# =====
_REQUEST_AUTH_INFO = "_kvmd_auth_info"


def _format_P(request: BaseRequest, *_, **__) -> str:  # type: ignore  # pylint: disable=invalid-name
    return (getattr(request, _REQUEST_AUTH_INFO, None) or "-")


AccessLogger._format_P = staticmethod(_format_P)  # type: ignore  # pylint: disable=protected-access


def set_request_auth_info(request: BaseRequest, info: str) -> None:
    setattr(request, _REQUEST_AUTH_INFO, info)


# =====
@dataclasses.dataclass(frozen=True)
class WsSession:
    wsr: WebSocketResponse
    kwargs: Dict[str, Any]

    def __str__(self) -> str:
        return f"WsSession(id={id(self)}, {self.kwargs})"

    async def send_event(self, event_type: str, event: Optional[Dict]) -> None:
        await send_ws_event(self.wsr, event_type, event)


class HttpServer:
    def __init__(self) -> None:
        self.__ws_heartbeat: Optional[float] = None
        self.__ws_handlers: Dict[str, Callable] = {}
        self.__ws_sessions: List[WsSession] = []
        self.__ws_sessions_lock = asyncio.Lock()

    def run(
        self,
        unix_path: str,
        unix_rm: bool,
        unix_mode: int,
        heartbeat: float,
        access_log_format: str,
    ) -> None:

        self.__ws_heartbeat = heartbeat

        if unix_rm and os.path.exists(unix_path):
            os.remove(unix_path)
        server_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        server_socket.bind(unix_path)
        if unix_mode:
            os.chmod(unix_path, unix_mode)

        run_app(
            sock=server_socket,
            app=self.__make_app(),
            shutdown_timeout=1,
            access_log_format=access_log_format,
            print=self.__run_app_print,
            # loop=asyncio.get_event_loop(),
        )

    # =====

    def _add_exposed(self, *objs: object) -> None:
        for obj in objs:
            for http_exposed in _get_exposed_http(obj):
                self.__add_exposed_http(http_exposed)
            for ws_exposed in _get_exposed_ws(obj):
                self.__add_exposed_ws(ws_exposed)

    def __add_exposed_http(self, exposed: HttpExposed) -> None:
        async def wrapper(request: Request) -> Response:
            try:
                await self._check_request_auth(exposed, request)
                return (await exposed.handler(request))
            except IsBusyError as err:
                return make_json_exception(err, 409)
            except (ValidatorError, OperationError) as err:
                return make_json_exception(err, 400)
            except HttpError as err:
                return make_json_exception(err)
        self.__app.router.add_route(exposed.method, exposed.path, wrapper)

    def __add_exposed_ws(self, exposed: WsExposed) -> None:
        self.__ws_handlers[exposed.event_type] = exposed.handler

    # =====

    @contextlib.asynccontextmanager
    async def _ws_session(self, request: Request, **kwargs: Any) -> AsyncGenerator[WsSession, None]:
        assert self.__ws_heartbeat is not None
        wsr = WebSocketResponse(heartbeat=self.__ws_heartbeat)
        await wsr.prepare(request)
        ws = WsSession(wsr, kwargs)

        async with self.__ws_sessions_lock:
            self.__ws_sessions.append(ws)
            get_logger(2).info("Registered new client session: %s; clients now: %d", ws, len(self.__ws_sessions))

        try:
            await self._on_ws_opened()
            yield ws
        finally:
            await aiotools.shield_fg(self.__close_ws(ws))

    async def _ws_loop(self, ws: WsSession) -> WebSocketResponse:
        logger = get_logger()
        async for msg in ws.wsr:
            if msg.type != WSMsgType.TEXT:
                break
            try:
                (event_type, event) = parse_ws_event(msg.data)
            except Exception as err:
                logger.error("Can't parse JSON event from websocket: %r", err)
            else:
                handler = self.__ws_handlers.get(event_type)
                if handler:
                    await handler(ws, event)
                else:
                    logger.error("Unknown websocket event: %r", msg.data)
        return ws.wsr

    async def _broadcast_ws_event(self, event_type: str, event: Optional[Dict]) -> None:
        if self.__ws_sessions:
            await asyncio.gather(*[
                ws.send_event(event_type, event)
                for ws in self.__ws_sessions
                if (
                    not ws.wsr.closed
                    and ws.wsr._req is not None  # pylint: disable=protected-access
                    and ws.wsr._req.transport is not None  # pylint: disable=protected-access
                )
            ], return_exceptions=True)

    async def _close_all_wss(self) -> bool:
        wss = self._get_wss()
        for ws in wss:
            await self.__close_ws(ws)
        return bool(wss)

    def _get_wss(self) -> List[WsSession]:
        return list(self.__ws_sessions)

    async def __close_ws(self, ws: WsSession) -> None:
        async with self.__ws_sessions_lock:
            try:
                self.__ws_sessions.remove(ws)
                get_logger(3).info("Removed client socket: %s; clients now: %d", ws, len(self.__ws_sessions))
                await ws.wsr.close()
            except Exception:
                pass
        await self._on_ws_closed()

    # =====

    async def _check_request_auth(self, exposed: HttpExposed, request: Request) -> None:
        pass

    async def _init_app(self) -> None:
        raise NotImplementedError

    async def _on_shutdown(self) -> None:
        pass

    async def _on_cleanup(self) -> None:
        pass

    async def _on_ws_opened(self) -> None:
        pass

    async def _on_ws_closed(self) -> None:
        pass

    # =====

    async def __make_app(self) -> Application:
        self.__app = Application(middlewares=[normalize_path_middleware(  # pylint: disable=attribute-defined-outside-init
            append_slash=False,
            remove_slash=True,
            merge_slashes=True,
        )])

        async def on_shutdown(_: Application) -> None:
            await self._on_shutdown()
        self.__app.on_shutdown.append(on_shutdown)

        async def on_cleanup(_: Application) -> None:
            await self._on_cleanup()
        self.__app.on_cleanup.append(on_cleanup)

        await self._init_app()
        return self.__app

    def __run_app_print(self, text: str) -> None:
        logger = get_logger(0)
        for line in text.strip().splitlines():
            logger.info(line.strip())


kvmd/apps/kvmd/sysunit.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2022  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


import types

from typing import Tuple
from typing import Type
from typing import Optional

import dbus
import dbus.exceptions

from ... import aiotools


# =====
class SystemdUnitInfo:
    def __init__(self) -> None:
        self.__bus: Optional[dbus.SystemBus] = None
        self.__manager: Optional[dbus.Interface] = None

    async def get_status(self, name: str) -> Tuple[bool, bool]:
        assert self.__bus is not None
        assert self.__manager is not None

        if not name.endswith(".service"):
            name += ".service"

        return await aiotools.run_async(self.__inner_get_status, name)

    def __inner_get_status(self, name: str) -> Tuple[bool, bool]:
        try:
            unit_p = self.__manager.GetUnit(name)
            unit = self.__bus.get_object("org.freedesktop.systemd1", unit_p)
            unit_props = dbus.Interface(unit, dbus_interface="org.freedesktop.DBus.Properties")
            started = (unit_props.Get("org.freedesktop.systemd1.Unit", "ActiveState") == "active")
        except dbus.exceptions.DBusException as err:
            if "NoSuchUnit" not in str(err):
                raise
            started = False
        enabled = (self.__manager.GetUnitFileState(name) in [
            "enabled",
            "enabled-runtime",
            "static",
            "indirect",
            "generated",
        ])
        return (enabled, started)

    async def open(self) -> None:
        await aiotools.run_async(self.__inner_open)

    def __inner_open(self) -> None:
        self.__bus = dbus.SystemBus()
        systemd = self.__bus.get_object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
        self.__manager = dbus.Interface(systemd, dbus_interface="org.freedesktop.systemd1.Manager")

    async def __aenter__(self) -> "SystemdUnitInfo":
        await self.open()
        return self

    async def close(self) -> None:
        try:
            if self.__bus is not None:
                await aiotools.run_async(self.__bus.close)
        except Exception:
            pass
        self.__manager = None
        self.__bus = None

    async def __aexit__(
        self,
        _exc_type: Type[BaseException],
        _exc: BaseException,
        _tb: types.TracebackType,
    ) -> None:

        await self.close()


kvmd/apps/kvmd/api/msd.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2022  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


import lzma
import time

from typing import Dict
from typing import AsyncGenerator
from typing import Optional
from typing import Union

import aiohttp

from aiohttp.web import Request
from aiohttp.web import Response
from aiohttp.web import StreamResponse

from ....logging import get_logger

from .... import aiotools
from .... import htclient

from ....htserver import exposed_http
from ....htserver import make_json_response
from ....htserver import make_json_exception
from ....htserver import start_streaming
from ....htserver import stream_json
from ....htserver import stream_json_exception

from ....plugins.msd import BaseMsd

from ....validators import check_string_in_list
from ....validators.basic import valid_bool
from ....validators.basic import valid_int_f0
from ....validators.basic import valid_float_f01
from ....validators.net import valid_url
from ....validators.kvm import valid_msd_image_name


# ======
class MsdApi:
    def __init__(self, msd: BaseMsd) -> None:
        self.__msd = msd

    # =====

    @exposed_http("GET", "/msd")
    async def __state_handler(self, _: Request) -> Response:
        return make_json_response(await self.__msd.get_state())

    @exposed_http("POST", "/msd/set_params")
    async def __set_params_handler(self, request: Request) -> Response:
        params = {
            key: validator(request.query.get(param))
            for (param, key, validator) in [
                ("image", "name", (lambda arg: str(arg).strip() and valid_msd_image_name(arg))),
                ("cdrom", "cdrom", valid_bool),
                ("rw", "rw", valid_bool),
            ]
            if request.query.get(param) is not None
        }
        await self.__msd.set_params(**params)  # type: ignore
        return make_json_response()

    @exposed_http("POST", "/msd/set_connected")
    async def __set_connected_handler(self, request: Request) -> Response:
        await self.__msd.set_connected(valid_bool(request.query.get("connected")))
        return make_json_response()

    # =====

    @exposed_http("GET", "/msd/read")
    async def __read_handler(self, request: Request) -> StreamResponse:
        name = valid_msd_image_name(request.query.get("image"))
        compressors = {
            "": ("", None),
            "none": ("", None),
            "lzma": (".xz", (lambda: lzma.LZMACompressor())),  # pylint: disable=unnecessary-lambda
        }

        try:
            zstandard = __import__("zstandard")
            compressors["zstandard"] = (".zst", (lambda: zstandard.ZstdCompressor().compressobj()))  # pylint: disable=unnecessary-lambda
        except Exception:
            pass

        (suffix, make_compressor) = compressors[check_string_in_list(
            arg=request.query.get("compress", ""),
            name="Compression mode",
            variants=set(compressors),
        )]

        async with self.__msd.read_image(name) as reader:
            if make_compressor is None:
                src = reader.read_chunked()
                size = reader.get_total_size()

            else:
                async def compressed() -> AsyncGenerator[bytes, None]:
                    assert make_compressor is not None
                    compressor = make_compressor()  # pylint: disable=not-callable
                    limit = reader.get_chunk_size()
                    buf = b""
                    try:
                        async for chunk in reader.read_chunked():
                            buf += await aiotools.run_async(compressor.compress, chunk)
                            if len(buf) >= limit:
                                yield buf
                                buf = b""
                    finally:
                        # Закрыть в любом случае
                        buf += await aiotools.run_async(compressor.flush)
                    if len(buf) > 0:
                        yield buf

                src = compressed()
                size = -1

            response = await start_streaming(request, "application/octet-stream", size, name + suffix)
            async for chunk in src:
                await response.write(chunk)
            return response

    # =====

    @exposed_http("POST", "/msd/write")
    async def __write_handler(self, request: Request) -> Response:
        name = valid_msd_image_name(request.query.get("image"))
        size = valid_int_f0(request.content_length)
        remove_incomplete = self.__get_remove_incomplete(request)
        written = 0
        async with self.__msd.write_image(name, size, remove_incomplete) as writer:
            chunk_size = writer.get_chunk_size()
            while True:
                chunk = await request.content.read(chunk_size)
                if not chunk:
                    break
                written = await writer.write_chunk(chunk)
        return make_json_response(self.__make_write_info(name, size, written))

    @exposed_http("POST", "/msd/write_remote")
    async def __write_remote_handler(self, request: Request) -> Union[Response, StreamResponse]:  # pylint: disable=too-many-locals
        url = valid_url(request.query.get("url"))
        insecure = valid_bool(request.query.get("insecure", False))
        timeout = valid_float_f01(request.query.get("timeout", 10.0))
        remove_incomplete = self.__get_remove_incomplete(request)

        name = ""
        size = written = 0
        response: Optional[StreamResponse] = None

        async def stream_write_info() -> None:
            assert response is not None
            await stream_json(response, self.__make_write_info(name, size, written))

        try:
            async with htclient.download(
                url=url,
                verify=(not insecure),
                timeout=timeout,
                read_timeout=(7 * 24 * 3600),
            ) as remote:

                name = str(request.query.get("image", "")).strip()
                if len(name) == 0:
                    name = htclient.get_filename(remote)
                name = valid_msd_image_name(name)

                size = valid_int_f0(remote.content_length)

                get_logger(0).info("Downloading image %r as %r to MSD ...", url, name)
                async with self.__msd.write_image(name, size, remove_incomplete) as writer:
                    chunk_size = writer.get_chunk_size()
                    response = await start_streaming(request, "application/x-ndjson")
                    await stream_write_info()
                    last_report_ts = 0
                    async for chunk in remote.content.iter_chunked(chunk_size):
                        written = await writer.write_chunk(chunk)
                        now = int(time.time())
                        if last_report_ts + 1 < now:
                            await stream_write_info()
                            last_report_ts = now

                await stream_write_info()
                return response

        except Exception as err:
            if response is not None:
                await stream_write_info()
                await stream_json_exception(response, err)
            elif isinstance(err, aiohttp.ClientError):
                return make_json_exception(err, 400)
            raise

    def __get_remove_incomplete(self, request: Request) -> Optional[bool]:
        flag: Optional[str] = request.query.get("remove_incomplete")
        return (valid_bool(flag) if flag is not None else None)

    def __make_write_info(self, name: str, size: int, written: int) -> Dict:
        return {"image": {"name": name, "size": size, "written": written}}

    # =====

    @exposed_http("POST", "/msd/remove")
    async def __remove_handler(self, request: Request) -> Response:
        await self.__msd.remove(valid_msd_image_name(request.query.get("image")))
        return make_json_response()

    @exposed_http("POST", "/msd/reset")
    async def __reset_handler(self, _: Request) -> Response:
        await self.__msd.reset()
        return make_json_response()
